server/src/main/java/org/opensearch/cluster/coordination/JoinHelper.java

```diff
@@ -343,11 +350,22 @@ public void handleResponse(Empty response) {
                         logger.debug("successfully joined {} with {}", destination, joinRequest);
                         lastFailedJoinAttempt.set(null);
                         onCompletion.run();
+                        if (isDecommissioned) {
+                            isDecommissioned = false;
+                            nodeCommissionedListener.onResponse(null);
+                        }
                     }
 
                     @Override
                     public void handleException(TransportException exp) {
                         pendingOutgoingJoins.remove(dedupKey);
+                        if (exp instanceof RemoteTransportException && (exp.getCause() instanceof NodeDecommissionedException)) {
+                            logger.info("local node is decommissioned. Will not be able to join the cluster");
+                            if (!isDecommissioned) {
+                                isDecommissioned = true;
+                                nodeCommissionedListener.onFailure(exp);
```

Comment:
nit : it doesn't look like a `Failure`, but an expected behavior. I would prefer just a `Executor` over `ActionListener`  for same reason . 

But overall it looks good.